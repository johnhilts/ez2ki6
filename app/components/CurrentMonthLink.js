import React from 'react';
import { Link } from 'react-router'
import * as dateUtils from '../util/dateutils';

export default function CurrentMonthLink(props) {
  let dayInfo = props.dayInfo;
  let monthName = dateUtils.getMonthName(dayInfo.month);
  return (
    <div style={{marginBottom: 15, fontSize: 'small', }}>
      <Link to={{pathname: 'calendar', query: {ym: dateUtils.getFormattedYearMonthByYearMonth(dayInfo.year, dayInfo.month) , },}}><h4>&gt;&gt;Return to {monthName}&lt;&lt;</h4></Link>
    </div>
  )
}
