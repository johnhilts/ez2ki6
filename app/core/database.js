export const firebaseConfig = {
  apiKey: 'AIzaSyDIoxipbDfUyTsZT_mCj-lShyNOfh_2ROg',
  authDomain: 'easynikkiroku.firebaseapp.com',
  databaseURL: 'https://easynikkiroku.firebaseio.com',
  storageBucket: '',
};

export function getUserRoot(owner) {
  return 'users/' + owner;
}
