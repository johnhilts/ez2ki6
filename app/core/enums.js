export const detailLevel = {day: 0, week: 1, month: 2, month_with_data: 3, };
