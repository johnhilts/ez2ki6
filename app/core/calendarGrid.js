import * as dateUtils from '../util/dateutils';
import * as enums from '../core/enums';

const saturday = 6;

export const getDayInfo = (dates, currentDate, isCurrentMonth, absoluteCellIndex) => {

  const matchesDate = dbDate => {return dbDate.year == currentDate.year && dbDate.month == currentDate.month && dbDate.day == currentDate.day;};

  let isToday = dateUtils.isToday(currentDate.year, currentDate.month, currentDate.day);
  let {year, month, day} = currentDate;
  let dayInfo = {year, month, day, absoluteIndex: absoluteCellIndex, isToday: isToday, isCurrentMonth: isCurrentMonth, };
  dayInfo.hasData = dates.some(matchesDate);
  if (dayInfo.hasData) {
    dayInfo.dateEntries = dates.filter(matchesDate);
  }

  return dayInfo;
}

export const getLastCalendarCellIndexFromStartOfCurrentMonth = (year, month, lastDay) => {
  let lastCurrentMonthWeekday = dateUtils.getWeekdayFromYearMonthDay(year, month, lastDay);
  return (
    lastCurrentMonthWeekday == saturday
    ? lastDay
    : lastDay + (saturday - lastCurrentMonthWeekday)
  )
}

export const buildMonthGrid = (dates, currentYearMonth) => {

  const getDayCellInfo = (currentDate, isCurrentMonth) => {
    return getDayInfo(dates, currentDate, isCurrentMonth, absoluteCellIndex++);
  }

  const daysInMonth = currentYearMonth.daysInMonth;
  let monthGrid = [];
  let absoluteCellIndex = 0;
  const currentYear = currentYearMonth.year;
  const currentMonth = currentYearMonth.month;
  const lastCalendarCellIndex = getLastCalendarCellIndexFromStartOfCurrentMonth(currentYear, currentMonth, daysInMonth);

  const processCalendarCell = (cellIndex) => {
    let day = cellIndex + 1;
    let date = dateUtils.getDateFromYearMonthDay(currentYear, currentMonth, day);

    const partOfWeekInPreviousMonth = (cellIndex) => {
      let previousMonthDate = dateUtils.addDays(currentYear, currentMonth, day, (weekday - cellIndex) * -1);
      return getDayCellInfo(previousMonthDate, false);
    }

    const weekInCurrentMonth = (cellIndex) => {
      return getDayCellInfo(date, true);
    }

    const partOfWeekInNextMonth = (cellIndex) => {
      return getDayCellInfo(date, false);
    }

    const daysInSameWeekPreviousMonth = (weekCellIndex) => {
      return day == 1 && weekday > weekCellIndex && weekCellIndex < saturday;
    }

    const daysInCurrentMonth = (weekCellIndex) => {
      return weekday == weekCellIndex && day <= daysInMonth;
    }

    const daysInSameWeekNextMonth = (weekCellIndex) => {
      return weekday == weekCellIndex && day > daysInMonth;
    }

    const weekLength = 7;
    let calendarCellsForWeek = [...Array(weekLength).keys()];
    let weekday = dateUtils.getWeekdayFromYearMonthDay(currentYear, currentMonth, day);

    monthGrid.push(...calendarCellsForWeek.filter(daysInSameWeekPreviousMonth).map(partOfWeekInPreviousMonth));
    monthGrid.push(...calendarCellsForWeek.filter(daysInCurrentMonth).map(weekInCurrentMonth));
    monthGrid.push(...calendarCellsForWeek.filter(daysInSameWeekNextMonth).map(partOfWeekInNextMonth));
  }

  let calendarCellsFromStartOfCurrentMonth = [...Array(lastCalendarCellIndex).keys()];
  calendarCellsFromStartOfCurrentMonth.map(processCalendarCell);

  return monthGrid;
}

export const setLinkDetailLevelForMonthView = (monthGrid) => {
  if (monthGrid.some(m => m.hasData)) {
    return enums.detailLevel.month_with_data
  }
  else {
    return enums.detailLevel.month
  }
}
