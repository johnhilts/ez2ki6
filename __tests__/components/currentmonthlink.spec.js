import React, { Component } from 'react';
import ReactTestRenderer from 'react-test-renderer';
import CurrentMonthLink from '../../app/components/CurrentMonthLink';
import * as calendarGrid from '../../app/core/calendarGrid';
import * as testHelper from '../../test/Helper/DateHelper';

describe('CurrentMonthLink component snapshot test', () => {

  it('Should compare the CurrentMonthLink component with a snapshot', () => {
    let dayInfo = {year: 2017, month: 4, };
    const component = ReactTestRenderer.create(
      <CurrentMonthLink dayInfo={calendarGrid.buildMonthGrid(testHelper.dates, testHelper.currentYearMonth)} />
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  })
});
