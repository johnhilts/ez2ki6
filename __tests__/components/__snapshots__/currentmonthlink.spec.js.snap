exports[`CurrentMonthLink component snapshot test Should compare the CurrentMonthLink component with a snapshot 1`] = `
<div
  style={
    Object {
      "fontSize": "small",
      "marginBottom": 15,
    }
  }>
  <a
    onClick={[Function]}
    style={Object {}}>
    <h4>
      >>Return to 
      Invalid date
      <<
    </h4>
  </a>
</div>
`;
